### PR TITLE
JAMES-3905 LDAP should allow per user base DN

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/data/data-ldap/src/test/resources/ldif-files/populate.ldif [hunk 1]
```diff
@@ -14,3 +18,12 @@ sn: james-user
 mail: james-user@james.org
 userPassword: secret
 description: James user
+
+dn: uid=bob, ou=whatever, dc=james,dc=org
```
Comment (Arsnael) [inline server/data/data-ldap/src/test/resources/ldif-files/populate.ldif:22]: My knowledge of ldap is limited but shouldn't it be more logical to have:

```suggestion
dn: uid=bob, ou=whatever, dc=extra,dc=org
```
instead?

File: server/data/data-ldap/src/test/resources/ldif-files/populate.ldif [hunk 2]
```diff
@@ -6,6 +6,10 @@ dn: ou=empty, dc=james,dc=org
 ou: empty
 objectClass: organizationalUnit
 
+dn: ou=whatever, dc=james,dc=org
```
Comment (Arsnael) [inline server/data/data-ldap/src/test/resources/ldif-files/populate.ldif:9]: idem here? dc=james => dc=extra, if domain name is extra.org?

File: server/data/data-ldap/src/test/resources/ldif-files/populate.ldif (hunk 1 above)
Comment (chibenwa) [reply server/data/data-ldap/src/test/resources/ldif-files/populate.ldif:22]: No, that's a limitation from dinkel container, I struggled a lot with it... OU needs to be based on the LDAP domain...

File: server/data/data-ldap/src/main/java/org/apache/james/user/ldap/LdapRepositoryConfiguration.java [hunk 3]
```diff
@@ -250,12 +275,17 @@ public static LdapRepositoryConfiguration from(HierarchicalConfiguration<Immutab
...
+
     private final boolean trustAllCerts;
 
+    private final ImmutableMap<Domain, String> perDomainBaseDN;
+
     private LdapRepositoryConfiguration(String ldapHost, String principal, String credentials, String userBase, String userIdAttribute,
                                         String userObjectClass, int connectionTimeout, int readTimeout,
                                         boolean supportsVirtualHosting, int poolSize, ReadOnlyLDAPGroupRestriction restriction, String filter,
-                                        Optional<String> administratorId, boolean trustAllCerts) throws ConfigurationException {
+
```
Comment (quantranhong1999) [inline server/data/data-ldap/src/main/java/org/apache/james/user/ldap/LdapRepositoryConfiguration.java:286]:
```suggestion
```

redundant line

**Review by quantranhong1999 — COMMENTED** [review]
nothing to add

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**